In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy.stats as sts
import statsmodels.api as sm
import warnings


from statsmodels.tsa.stattools import acf
from sklearn.preprocessing import normalize
from sklearn.preprocessing import scale
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from scipy.stats.mstats import winsorize
from scipy.stats import jarque_bera
from scipy.stats import normaltest
from scipy.stats import bartlett
from scipy.stats import levene
from sqlalchemy import create_engine
from sklearn import linear_model

warnings.filterwarnings('ignore')

In [2]:
hava=pd.read_csv(".\\WeatherHistory.csv")

In [9]:
hava["fark"]=hava["Temperature (C)"]-hava["Apparent Temperature (C)"]

In [10]:
hava.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary      fark  
0  Partly cloudy throughout the day.  2.083333  
1  Partly cloudy throughout the day.  2.127778  
2  Partly cloudy throughout the day.  0.000000  
3  Partly cloudy throughout the day.  2.344444  
4  Partly cloudy throughout the day.  1.777778

In [11]:
Y=hava["fark"]
X=hava[["Humidity", "Wind Speed (km/h)"]]

In [12]:
X=sm.add_constant(X)

sonuc=sm.OLS(Y, X).fit()
sonuc.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   fark   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Mon, 15 Apr 2019   Prob (F-statistic):               0.00
Time:                        08:22:58   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

##### MODELİMİZ:

fark= -2.4381 + 3.0292 *nem + 0.1193 *rüzgar_hızı

Rüzgâr hızının nemden daha fazla etkili olmasını beklerdim. P değerleri 0 (sıfır) olduğu için modelimizin katsayıları anlamlıdır. Sıcaklık farkı; nem 1 birim arttığında 3.0292 derece, rüzgâr 1 birim arttığında 0.1193 derece artmaktadır. İlişki eş yönlüdür.

In [13]:
hava["Hum_Wind"]=hava["Humidity"]*hava["Wind Speed (km/h)"]

X=hava[["Humidity", "Wind Speed (km/h)", "Hum_Wind"]]

X=sm.add_constant(X)

sonuc=sm.OLS(Y, X).fit()
sonuc.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   fark   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Mon, 15 Apr 2019   Prob (F-statistic):               0.00
Time:                        08:37:16   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Hum_Wind              0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

##### YENİ MODELİMİZ:

fark= -0.0839 - 0.1775 *nem - 0.0905 *rüzgar_hızı + 0.2971 *nem_rüzgâr_etkileşimi

P değerleri 0.05'ten küçük olduğu için katsayılar anlamlıdır. Rüzgâr ve nemin katsayıları işaret değiştirdi. Etkileşim katsayısı daha büyük bir değer aldı. İlk modelde, rüzgâr ve nem sıcaklık farkını artırıcı bir rol oynarken, yeni modelimizde azaltıcı bir rol oynamaya başladı. Nem etkisi halâ rüzgarı etkisinden fazla.